[GTK] Use DMABuf and WebKit IPC for rendering instead of wpe/x11

carlosgcampos opened the pull request:
#### d2424c0fc67f6a46921edb52874720bb99c0c699
<pre>
[GTK] Use DMABuf and WebKit IPC for rendering instead of wpe/x11
<a href="https://bugs.webkit.org/show_bug.cgi?id=252988">https://bugs.webkit.org/show_bug.cgi?id=252988</a>

Reviewed by Žan Doberšek.

Instead of the wpe rendered, which uses a nested wayland compositor, and
x11, which uses a redirected xcomposite window, we can make the web
process render always into a pbuffer using mesa surfaceless platform and
export the rendered texture using DMAbuf to be consumed by the UI
process. This way the rendering is independent from the platform
(wayland/x11), and in the UI process we just import the DMABuf into a
texture that we pass to GTK for rendering into the web view widget. If
under X11 GTK uses GLX instead of EGL we manually import the DMABuf
using libgbm to create a cairo surface that we can pass to GTK without
using GL.

* Source/WebCore/PlatformGTK.cmake:
* Source/WebCore/SourcesGTK.txt:
* Source/WebCore/platform/graphics/PlatformDisplay.cpp:
(WebCore::PlatformDisplay::initializeEGLDisplay):
* Source/WebCore/platform/graphics/PlatformDisplay.h:
* Source/WebCore/platform/graphics/egl/GLContextEGL.cpp:
(WebCore::GLContextEGL::getEGLConfig):
(WebCore::GLContextEGL::createWindowContext):
(WebCore::GLContextEGL::createPbufferContext):
(WebCore::GLContextEGL::createSurfacelessContext):
(WebCore::GLContextEGL::createContext):
(WebCore::GLContextEGL::createSharingContext):
* Source/WebCore/platform/graphics/egl/GLContextEGL.h:
* Source/WebCore/platform/graphics/egl/GLContextEGLLibWPE.cpp:
(WebCore::GLContextEGL::createWPEContext):
* Source/WebCore/platform/graphics/egl/GLContextEGLWayland.cpp:
(WebCore::GLContextEGL::createWaylandContext):
* Source/WebCore/platform/graphics/egl/GLContextEGLX11.cpp:
(WebCore::GLContextEGL::createPixmapContext):
* Source/WebCore/platform/graphics/egl/PlatformDisplayHeadless.cpp: Added.
(WebCore::PlatformDisplayHeadless::create):
(WebCore::PlatformDisplayHeadless::PlatformDisplayHeadless):
(WebCore::PlatformDisplayHeadless::~PlatformDisplayHeadless):
* Source/WebCore/platform/graphics/egl/PlatformDisplayHeadless.h: Added.
* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/CoordinatedGraphics/threadedcompositor/ThreadedCompositor.cpp:
(WebKit::ThreadedCompositor::createGLContext):
(WebKit::ThreadedCompositor::invalidate):
* Source/WebKit/Shared/CoordinatedGraphics/threadedcompositor/ThreadedCompositor.h:
* Source/WebKit/Shared/WebProcessCreationParameters.cpp:
(WebKit::WebProcessCreationParameters::encode const):
(WebKit::WebProcessCreationParameters::decode):
* Source/WebKit/Shared/WebProcessCreationParameters.h:
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/UIProcess/API/glib/WebKitProtocolHandler.cpp:
(WebKit::WebKitProtocolHandler::handleGPU):
* Source/WebKit/UIProcess/glib/WebProcessPoolGLib.cpp:
(WebKit::WebProcessPool::platformInitializeWebProcess):
* Source/WebKit/UIProcess/gtk/AcceleratedBackingStore.cpp:
(WebKit::AcceleratedBackingStore::checkRequirements):
(WebKit::AcceleratedBackingStore::create):
* Source/WebKit/UIProcess/gtk/AcceleratedBackingStoreDMABuf.cpp: Added.
(WebKit::gtkGLContextIsEGL):
(WebKit::AcceleratedBackingStoreDMABuf::checkRequirements):
(WebKit::AcceleratedBackingStoreDMABuf::create):
(WebKit::AcceleratedBackingStoreDMABuf::AcceleratedBackingStoreDMABuf):
(WebKit::AcceleratedBackingStoreDMABuf::~AcceleratedBackingStoreDMABuf):
(WebKit::AcceleratedBackingStoreDMABuf::RenderSource::RenderSource):
(WebKit::AcceleratedBackingStoreDMABuf::Texture::Texture):
(WebKit::AcceleratedBackingStoreDMABuf::Texture::~Texture):
(WebKit::AcceleratedBackingStoreDMABuf::Texture::swap):
(WebKit::AcceleratedBackingStoreDMABuf::Texture::snapshot const):
(WebKit::AcceleratedBackingStoreDMABuf::Texture::paint const):
(WebKit::AcceleratedBackingStoreDMABuf::Surface::Surface):
(WebKit::AcceleratedBackingStoreDMABuf::Surface::~Surface):
(WebKit::AcceleratedBackingStoreDMABuf::Surface::map const):
(WebKit::AcceleratedBackingStoreDMABuf::Surface::swap):
(WebKit::AcceleratedBackingStoreDMABuf::Surface::snapshot const):
(WebKit::AcceleratedBackingStoreDMABuf::Surface::paint const):
(WebKit::AcceleratedBackingStoreDMABuf::configure):
(WebKit::AcceleratedBackingStoreDMABuf::frame):
(WebKit::AcceleratedBackingStoreDMABuf::unrealize):
(WebKit::AcceleratedBackingStoreDMABuf::ensureGLContext):
(WebKit::AcceleratedBackingStoreDMABuf::makeContextCurrent):
(WebKit::AcceleratedBackingStoreDMABuf::update):
(WebKit::AcceleratedBackingStoreDMABuf::snapshot):
(WebKit::AcceleratedBackingStoreDMABuf::paint):
* Source/WebKit/UIProcess/gtk/AcceleratedBackingStoreDMABuf.h: Added.
* Source/WebKit/UIProcess/gtk/AcceleratedBackingStoreDMABuf.messages.in: Added.
* Source/WebKit/WebProcess/WebPage/AcceleratedSurface.cpp:
(WebKit::AcceleratedSurface::create):
* Source/WebKit/WebProcess/WebPage/AcceleratedSurface.h:
(WebKit::AcceleratedSurface::didCreateGLContext):
(WebKit::AcceleratedSurface::willDestroyGLContext):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/LayerTreeHost.cpp:
(WebKit::LayerTreeHost::didCreateGLContext):
(WebKit::LayerTreeHost::willDestroyGLContext):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/LayerTreeHost.h:
* Source/WebKit/WebProcess/WebPage/gtk/AcceleratedSurfaceDMABuf.cpp: Added.
(WebKit::AcceleratedSurfaceDMABuf::create):
(WebKit::AcceleratedSurfaceDMABuf::AcceleratedSurfaceDMABuf):
(WebKit::AcceleratedSurfaceDMABuf::~AcceleratedSurfaceDMABuf):
(WebKit::AcceleratedSurfaceDMABuf::didCreateGLContext):
(WebKit::AcceleratedSurfaceDMABuf::willDestroyGLContext):
(WebKit::AcceleratedSurfaceDMABuf::surfaceID const):
(WebKit::AcceleratedSurfaceDMABuf::clientResize):
(WebKit::AcceleratedSurfaceDMABuf::willRenderFrame):
(WebKit::AcceleratedSurfaceDMABuf::didRenderFrame):
* Source/WebKit/WebProcess/WebPage/gtk/AcceleratedSurfaceDMABuf.h: Added.
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/glib/WebProcessGLib.cpp:
(WebKit::WebProcess::platformInitializeWebProcess):

Canonical link: <a href="https://commits.webkit.org/262210@main">https://commits.webkit.org/262210@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0e0be399d0f3e1700e14e3273ac18a78bd044854

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/908 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/932 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/968 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/1302 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/803 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/895 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/982 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/1017 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/1007 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/915 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/853 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/853 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/1240 "Built successfully") | 
| [❌ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/935 "webkitpy-tests (failure)") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/904 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/850 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/847 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/822 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/877 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/1875 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/863 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/831 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/810 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/861 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/877 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/94 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebCore/platform/graphics/egl/GLContextEGL.cpp:
```diff
@@ -328,6 +330,15 @@ std::unique_ptr<GLContextEGL> GLContextEGL::createContext(GLNativeWindowType win
+        auto context = createPbufferContext(platformDisplay, eglSharingContext);
+        if (!context) {
+            RELEASE_LOG_INFO(Compositing, "Could not create Pbuffer context: %s.", lastErrorString());
+            WTFLogAlways("Could not create EGL context.");
```
Why not log the error message with WTFLogAlways? The release log stuff is useful for developers but users aren't going to provide it in bug reports and we probably won't remember to ask for it. More info in stderr is helpful IMO.

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebContext.cpp:
```diff
@@ -426,7 +426,7 @@ static void webkitWebContextConstructed(GObject* object)
     configuration.setUsesWebProcessCache(true);
 #if PLATFORM(GTK) && !USE(GTK4)
-    configuration.setProcessSwapsOnNavigation(priv->psonEnabled);
+    configuration.setProcessSwapsOnNavigation(false/*priv->psonEnabled*/);
```
Opening this conversation just to ensure there's no chance we'll forget this

mcatanzaro commented on Source/WebKit/UIProcess/gtk/AcceleratedBackingStoreDMABuf.cpp:
```diff
@@ -0,0 +1,266 @@
+        available = eglExtensions.KHR_image_base
+            && eglExtensions.KHR_surfaceless_context
+            && eglExtensions.EXT_image_dma_buf_import
+            && eglExtensions.MESA_image_dma_buf_export;
```
So I see MESA and assume this won't be available on NVIDIA proprietary graphics or other proprietary graphics drivers. Will it be available to everyone using mesa? Any example configurations that might result in getting the fallback path?

aperezdc replied:
The `MESA` bit in the name indicates where the extension originated, but does not prevent other drivers to implement it:

```
% strings libEGL_nvidia.so.525.89.02 | grep -i dmabuf      
eglQueryDmaBufFormatsEXT
eglQueryDmaBufModifiersEXT
eglExportDMABUFImageMESA
eglExportDMABUFImageQueryMESA
%
```

carlosgcampos replied:
It also requires `EGL_MESA_platform_surfaceless`. Situation is:
- `EGL_MESA_platform_surfaceless`
  - Supported in mesa since 2016 (don't know the exact version, it could be 12.x)
  - Enabled by default since mesa 17.1
  - Compiled unconditionally since mesa 20.2
- `EGL_MESA_image_dma_buf_export`
  - Initial support was added in 2014 (don't know the exact version, it cold be 10.x)
  - No compile option, so it's always avaialable

mcatanzaro commented on Source/WebKit/UIProcess/gtk/AcceleratedBackingStoreDMABuf.cpp:
```diff
@@ -0,0 +1,266 @@
+    }
+
+    g_warning("GDK is not able to create a GL context, falling back to ... (slow!): %s", error->message);
+    // FIXME: implement fallback?
```
Another placeholder discussion just to ensure we don't forget the FIXMEs

carlosgcampos replied:
So, the question here is whether we really want to keep using hardware acceleration when GTK can't. I think we could try to ensure GTK will be able to create a context in checkRequirements() (or even HardwareAccelerationManager) and simply disable hw acceleration if that fails.

zdobersek commented on Source/WebKit/WebProcess/WebPage/gtk/AcceleratedSurfaceDMABuf.cpp:
```diff
@@ -0,0 +1,134 @@
+    int fd, fourcc;
+    EGLint offset, stride;
+    eglExportDMABUFImageQueryMESA(eglDisplay, image, &fourcc, nullptr, nullptr);
+    eglExportDMABUFImageMESA(eglDisplay, image, &fd, &stride, &offset);
```
The extension itself is not widely supported. There's also no control over the parameters of the underlying dmabuf, meaning the internal implementation can select some combination of the format and modifier values that might cause problems during importing on the other side, because e.g. modifier information is not available or the GTK-controlled context ends up in a configuration where the import doesn't work.

Current approach also implements rendering into a single dmabuf, which is then sampled from in the UIProcess. This can cause problems if WebProcess renders into the dmabuf as it's being sampled in the UIProcess. Simple double-buffering with commit-release mechanics would work around that. Better yet, a GBMBufferSwapchain can be used similarly to how it's used in GraphicsContextGLGBM.

This approach was already implemented in WPEBackend-fdo, as the dmabuf-pooling variant. The rendering backend implementation is here:
https://github.com/Igalia/WPEBackend-fdo/blob/master/src/egl-client-dmabuf-pool.cpp

zdobersek commented on Source/WebKit/WebProcess/WebPage/gtk/AcceleratedSurfaceDMABuf.cpp:
```diff
@@ -0,0 +1,191 @@
+        return;
+
+    glFlush();
+    m_webPage.sendWithAsyncReply(Messages::AcceleratedBackingStoreDMABuf::Frame(), [this, weakThis = WeakPtr { *this }, runLoop = Ref { RunLoop::current() }]() mutable {
```
This is sending IPC messages through WebPage on the composition thread, which is problematic.

carlosgcampos replied:
What exactly is problematic? I thought it was thread safe.

zdobersek replied:
WebPage isn't usable across different threads. Currently these calls go through the IPC::MessageSender interface and the underlying Connection retrieval and the message-sending act end up working fine. But instead of relying on that, it would be easier on the mind if you retain a Connection ref and use that.

WebSWContextManagerConnection does that. It also runs a WorkQueue on which it receives the IPC messages, which is something we'd want here, but the composition thread is run through a RunLoop. So if desired, in the future a sideline WorkQueue would be receiving the messages and dispatching them on the composition thread.
https://github.com/WebKit/WebKit/blob/main/Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.cpp#L118

carlosgcampos replied:
Current version of the patch uses `WebProcess::singleton().parentProcessConnection()` instead. AcceleratedSurfaceDMABuf is not a message receiver, it only receives replies to messages with async reply, can we use a work queue in that case too?

zdobersek replied:
This has been changing recently. But using this parent-process connection, I think the async reply handling is done on the main thread. If there was a custom connection, it would  be possible to specify dispatching to occur on some different thread (see `Connection::open()`).

If `AcceleratedSurfaceDMABuf` is made a message receiver, then a specific FunctionDispatcher implementor can be tied to handling those messages, via `Connection::addMessageReceiver()`. A custom implementation of that interface could push message dispatches onto the composition thread.

zdobersek commented on Source/WebKit/WebProcess/WebPage/gtk/AcceleratedSurfaceDMABuf.cpp:
```diff
@@ -0,0 +1,191 @@
+    if (size.isEmpty())
+        return;
+
+    auto* backObject = gbm_bo_create(WebCore::GBMDevice::singleton().device(), size.width(), size.height(), uint32_t(DMABufFormat::FourCC::ARGB8888), 0);
```
Why not use `GBMBufferSwapchain` and pass `DMABufObject`s over IPC?

carlosgcampos replied:
I found it a bit complicated to use, two buffers should be enough and we don't need the eventfd, so we can save file descriptors open.

zdobersek replied:
I think it will be needed, at least for the general use case, but it can be adopted when that's taken up.

zdobersek commented on Source/WebKit/WebProcess/WebPage/gtk/AcceleratedSurfaceDMABuf.cpp:
```diff
@@ -0,0 +1,191 @@
+    glBindTexture(GL_TEXTURE_2D, m_texture);
+    glEGLImageTargetTexture2DOES(GL_TEXTURE_2D, m_backImage);
+    glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D, m_texture, 0);
+    glBindFramebuffer(GL_FRAMEBUFFER, m_fbo);
```
The order of these two has to be swapped.

carlosgcampos replied:
Ok.

zdobersek commented on Source/WebKit/WebProcess/WebPage/gtk/AcceleratedSurfaceDMABuf.cpp:
```diff
@@ -0,0 +1,191 @@
+        return;
+    }
+
+    m_webPage.send(Messages::AcceleratedBackingStoreDMABuf::Configure(WTFMove(backFD), WTFMove(frontFD), gbm_bo_get_format(backObject), gbm_bo_get_offset(backObject, 0), gbm_bo_get_stride(backObject), size));
```
This is sending IPC messages through WebPage on the composition thread, which is problematic.

zdobersek commented on Source/WebKit/WebProcess/WebPage/gtk/AcceleratedSurfaceDMABuf.cpp:
```diff
@@ -0,0 +1,191 @@
+
+    glBindTexture(GL_TEXTURE_2D, m_texture);
+    glEGLImageTargetTexture2DOES(GL_TEXTURE_2D, m_backImage);
+    glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D, m_texture, 0);
```
This should also handle the depth and the stencil attachments, both are needed by TextureMapper.

carlosgcampos replied:
I need hlp with that because I don't know what it is.

carlosgcampos replied:
I also added the depth and stencil attachments handling, hopefully the right way.

zdobersek replied:
Yes.

zdobersek commented on Source/WebCore/platform/graphics/PlatformDisplay.h:
```diff
@@ -78,6 +78,9 @@ class PlatformDisplay {
         WPE,
+#endif
+#if USE(EGL)
+        Headless,
```
Maybe `Surfaceless` would be better suitable, but it's not that important.

zdobersek commented on Source/WebCore/platform/graphics/egl/GLContextEGL.cpp:
```diff
@@ -125,18 +125,24 @@ bool GLContextEGL::getEGLConfig(EGLDisplay display, EGLConfig* config, EGLSurfac
     switch (surfaceType) {
+    case GLContextEGL::Surfaceless:
+        if (platformDisplay.type() == PlatformDisplay::Type::Headless)
+            attributeList[13] = EGL_PBUFFER_BIT;
```
From https://registry.khronos.org/EGL/extensions/MESA/EGL_MESA_platform_surfaceless.txt
```

    The surfaceless platform imposes no platform-specific restrictions on the
    creation of pbuffers, as eglCreatePbufferSurface has no native surface
    parameter.  Specifically, if the EGLDisplay advertises an EGLConfig whose
    EGL_SURFACE_TYPE attribute contains EGL_PBUFFER_BIT, then the EGLDisplay
    permits the creation of pbuffers with that config.

```

Do we end up using  pbuffers, or do we avoid them due to all the surfaceless extensions? If not used, then this attribute probably wouldn't need to be included in the list.

carlosgcampos replied:
I don't know if pbuffers are internally used or not, but I think the flag needs to be present for the config to match. See eglInfo output, the surfaceless platform only shows configs with pb surfaces supported.

zdobersek replied:
I think removing the attribute would still match the config. But this would also affect other users of GLContextEGL who might want to leverage pbuffers, so let's keep it in place for now.

zdobersek commented on Source/WebKit/UIProcess/gtk/AcceleratedBackingStoreDMABuf.cpp:
```diff
@@ -0,0 +1,438 @@
+
+    uint32_t mapStride = 0;
+    void* mapData = nullptr;
+    void* map = gbm_bo_map(buffer, 0, 0, static_cast<uint32_t>(m_size.width()), static_cast<uint32_t>(m_size.height()), GBM_BO_TRANSFER_READ, &mapStride, &mapData);
```
It's not great keeping a GPU memory mapping alive. I've had problems around simpler usage in GStreamer. Just immediate copying of that memory into a fastmalloced buffer and then unmapping would be safer.

carlosgcampos replied:
What kind of issues? I assumed it's ok as long as the web process doesn't change the buffer, which shouldn't happen.

zdobersek replied:
It was impossible to keep two mappings alive and stable at the same time.
http://webkit.org/b/253244

carlosgcampos replied:
Note that we always have only one buffer mapped here, the front one. When we swap buffers the previous one is unmapped and the new one mapped.

zdobersek replied:
We can address this later.

zdobersek commented on Source/WebKit/UIProcess/gtk/AcceleratedBackingStoreDMABuf.cpp:
```diff
@@ -0,0 +1,438 @@
+
+#if USE(GTK4)
+    glEGLImageTargetTexture2DOES(GL_TEXTURE_2D, m_backImage);
+    m_texture[0] = adoptGRef(gdk_gl_texture_new(m_context.get(), m_textureID, m_size.width(), m_size.height(), nullptr, nullptr));
+    m_texture[1] = adoptGRef(gdk_gl_texture_new(m_context.get(), m_textureID, m_size.width(), m_size.height(), nullptr, nullptr));
```
What's the reason for two separate GdkTexture objects, when both are using the same GL texture internally? As soon as you change the underlying EGLImage for that GL texture, both of these will be affected.

carlosgcampos replied:
Because of the way gdk works internally, if you pass a different GdkTexture object even if for the same texture, it makes a diff to get the changed region. That's whay I understood, see https://bugs.webkit.org/show_bug.cgi?id=252648

carlosgcampos committed:
[GTK] Use DMABuf and WebKit IPC for rendering instead of wpe/x11
https://bugs.webkit.org/show_bug.cgi?id=252988

Reviewed by Žan Doberšek.

Instead of the wpe rendered, which uses a nested wayland compositor, and
x11, which uses a redirected xcomposite window, we can make the web
process render always into a pbuffer using mesa surfaceless platform and
export the rendered texture using DMAbuf to be consumed by the UI
process. This way the rendering is independent from the platform
(wayland/x11), and in the UI process we just import the DMABuf into a
texture that we pass to GTK for rendering into the web view widget. If
under X11 GTK uses GLX instead of EGL we manually import the DMABuf
using libgbm to create a cairo surface that we can pass to GTK without
using GL.

* Source/WebCore/PlatformGTK.cmake:
* Source/WebCore/SourcesGTK.txt:
* Source/WebCore/platform/graphics/PlatformDisplay.cpp:
(WebCore::PlatformDisplay::initializeEGLDisplay):
* Source/WebCore/platform/graphics/PlatformDisplay.h:
* Source/WebCore/platform/graphics/egl/GLContextEGL.cpp:
(WebCore::GLContextEGL::getEGLConfig):
(WebCore::GLContextEGL::createWindowContext):
(WebCore::GLContextEGL::createPbufferContext):
(WebCore::GLContextEGL::createSurfacelessContext):
(WebCore::GLContextEGL::createContext):
(WebCore::GLContextEGL::createSharingContext):
* Source/WebCore/platform/graphics/egl/GLContextEGL.h:
* Source/WebCore/platform/graphics/egl/GLContextEGLLibWPE.cpp:
(WebCore::GLContextEGL::createWPEContext):
* Source/WebCore/platform/graphics/egl/GLContextEGLWayland.cpp:
(WebCore::GLContextEGL::createWaylandContext):
* Source/WebCore/platform/graphics/egl/GLContextEGLX11.cpp:
(WebCore::GLContextEGL::createPixmapContext):
* Source/WebCore/platform/graphics/egl/PlatformDisplayHeadless.cpp: Added.
(WebCore::PlatformDisplayHeadless::create):
(WebCore::PlatformDisplayHeadless::PlatformDisplayHeadless):
(WebCore::PlatformDisplayHeadless::~PlatformDisplayHeadless):
* Source/WebCore/platform/graphics/egl/PlatformDisplayHeadless.h: Added.
* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/CoordinatedGraphics/threadedcompositor/ThreadedCompositor.cpp:
(WebKit::ThreadedCompositor::createGLContext):
(WebKit::ThreadedCompositor::invalidate):
* Source/WebKit/Shared/CoordinatedGraphics/threadedcompositor/ThreadedCompositor.h:
* Source/WebKit/Shared/WebProcessCreationParameters.cpp:
(WebKit::WebProcessCreationParameters::encode const):
(WebKit::WebProcessCreationParameters::decode):
* Source/WebKit/Shared/WebProcessCreationParameters.h:
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/UIProcess/API/glib/WebKitProtocolHandler.cpp:
(WebKit::WebKitProtocolHandler::handleGPU):
* Source/WebKit/UIProcess/glib/WebProcessPoolGLib.cpp:
(WebKit::WebProcessPool::platformInitializeWebProcess):
* Source/WebKit/UIProcess/gtk/AcceleratedBackingStore.cpp:
(WebKit::AcceleratedBackingStore::checkRequirements):
(WebKit::AcceleratedBackingStore::create):
* Source/WebKit/UIProcess/gtk/AcceleratedBackingStoreDMABuf.cpp: Added.
(WebKit::gtkGLContextIsEGL):
(WebKit::AcceleratedBackingStoreDMABuf::checkRequirements):
(WebKit::AcceleratedBackingStoreDMABuf::create):
(WebKit::AcceleratedBackingStoreDMABuf::AcceleratedBackingStoreDMABuf):
(WebKit::AcceleratedBackingStoreDMABuf::~AcceleratedBackingStoreDMABuf):
(WebKit::AcceleratedBackingStoreDMABuf::RenderSource::RenderSource):
(WebKit::AcceleratedBackingStoreDMABuf::Texture::Texture):
(WebKit::AcceleratedBackingStoreDMABuf::Texture::~Texture):
(WebKit::AcceleratedBackingStoreDMABuf::Texture::swap):
(WebKit::AcceleratedBackingStoreDMABuf::Texture::snapshot const):
(WebKit::AcceleratedBackingStoreDMABuf::Texture::paint const):
(WebKit::AcceleratedBackingStoreDMABuf::Surface::Surface):
(WebKit::AcceleratedBackingStoreDMABuf::Surface::~Surface):
(WebKit::AcceleratedBackingStoreDMABuf::Surface::map const):
(WebKit::AcceleratedBackingStoreDMABuf::Surface::swap):
(WebKit::AcceleratedBackingStoreDMABuf::Surface::snapshot const):
(WebKit::AcceleratedBackingStoreDMABuf::Surface::paint const):
(WebKit::AcceleratedBackingStoreDMABuf::configure):
(WebKit::AcceleratedBackingStoreDMABuf::frame):
(WebKit::AcceleratedBackingStoreDMABuf::unrealize):
(WebKit::AcceleratedBackingStoreDMABuf::ensureGLContext):
(WebKit::AcceleratedBackingStoreDMABuf::makeContextCurrent):
(WebKit::AcceleratedBackingStoreDMABuf::update):
(WebKit::AcceleratedBackingStoreDMABuf::snapshot):
(WebKit::AcceleratedBackingStoreDMABuf::paint):
* Source/WebKit/UIProcess/gtk/AcceleratedBackingStoreDMABuf.h: Added.
* Source/WebKit/UIProcess/gtk/AcceleratedBackingStoreDMABuf.messages.in: Added.
* Source/WebKit/WebProcess/WebPage/AcceleratedSurface.cpp:
(WebKit::AcceleratedSurface::create):
* Source/WebKit/WebProcess/WebPage/AcceleratedSurface.h:
(WebKit::AcceleratedSurface::didCreateGLContext):
(WebKit::AcceleratedSurface::willDestroyGLContext):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/LayerTreeHost.cpp:
(WebKit::LayerTreeHost::didCreateGLContext):
(WebKit::LayerTreeHost::willDestroyGLContext):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/LayerTreeHost.h:
* Source/WebKit/WebProcess/WebPage/gtk/AcceleratedSurfaceDMABuf.cpp: Added.
(WebKit::AcceleratedSurfaceDMABuf::create):
(WebKit::AcceleratedSurfaceDMABuf::AcceleratedSurfaceDMABuf):
(WebKit::AcceleratedSurfaceDMABuf::~AcceleratedSurfaceDMABuf):
(WebKit::AcceleratedSurfaceDMABuf::didCreateGLContext):
(WebKit::AcceleratedSurfaceDMABuf::willDestroyGLContext):
(WebKit::AcceleratedSurfaceDMABuf::surfaceID const):
(WebKit::AcceleratedSurfaceDMABuf::clientResize):
(WebKit::AcceleratedSurfaceDMABuf::willRenderFrame):
(WebKit::AcceleratedSurfaceDMABuf::didRenderFrame):
* Source/WebKit/WebProcess/WebPage/gtk/AcceleratedSurfaceDMABuf.h: Added.
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/glib/WebProcessGLib.cpp:
(WebKit::WebProcess::platformInitializeWebProcess):

Canonical link: https://commits.webkit.org/262210@main
